### DISCUSSION TITLE
Calculate NEXTAUTH_URL based on incoming request

### DISCUSSION THREAD

**Discussion by buehler** [nextauthjs/next-auth#4517] [Ideas]
### Description 📓

Setting the NEXTAUTH_URL environment variable seems unnecessary in certain environments. For example, when running a next app on google cloud run, the URL of the application is only available _after_ the first deployment. Therefore, to create preview deployments, one needs to deploy an application twice. First, without the URL env, and after the url is available, a second time with the NEXTAUTH url var set.

As I'm aware of, other frameworks like ASP.Net do calculate the current host based on the incoming request. Would it be possible to do the same for next auth as well?

The calculation of ASP.net even works with proxies since the framework does respect X-Forwarded-For headers.

Then, one could remove the need of this variable altogether or just use it to overwrite the calculation for good.

### How to reproduce ☕️

Create a nextjs app with next auth (let's say a google login), and try to deploy it on google cloud run. 

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment]
This is currently only supported when deploying to Vercel. We cannot blindly trust request headers. Vercel overrides the correct ones in their proxy so it's safe to rely on it. Other proxies might not do that so we have to be careful and let the user be intentional about trusting headers. 

An environment variable or other methods are necessary.

Currently we detect if `VERCEL` exists and then trust the host.

See:
https://next-auth.js.org/deployment#vercel

https://github.com/nextauthjs/next-auth/blob/8b6d2e39721653f3c548ab593273c7da919722f2/packages/next-auth/src/next/utils.ts#L18-L23

**Reply by buehler** [reply]
Why do you only trust the Vercel host? I'm curious because other frameworks do exactly that. It's the programmer's responsibility to configure trust settings.

As I mentioned, in ASP.Net, one can configure trusted proxies, trusted networks and trusted headers. Then, the app is able to inherit the redirect information without the need for an environment variable.

As a sidenote: if I understand your code snippet correctly, I _could_ just set the vercel env var and then it would blindly trust the forwarded host, right?

**Comment by buehler** [comment]
anything? :-(

**Comment by DoodlesEpic** [comment]
Funnily enough the solution to the problem if you don't know the host in advance, or if it changes ports (for always on deployments for instance) is to just set `VERCEL=1` in your environment variables. The code path @balazsorban44 has put here shows that it essentially just trusts the host if that environment variable is set, which is ridiculous. Either this is pure ignore/incompetence in which case I'm not sure I want to trust this library, or it is a way to make people feel nudged to deploy to Vercel, in which case I hope this gets changed soon. The cookies implementation already does not need to use the `NEXTAUTH_URL` envvar and they have a comment saying `TODO: Remove this` at the definition of that function, so hopefully it's just a bit of tech debt.

The worst part of it for me is that setting it to `https://127.0.0.1:10000/api/auth/session` also works, and that is the URL which is printed in the console by the library. Trusting localhost is truly useless, so again I hope this gets changed soon.

**Reply by balazsorban44** [reply]
Currently undocumented, but a non-Vercel specific variable is also possible now.

https://github.com/nextauthjs/next-auth/blob/f12b5273006f4b3a6408c966b34349a12a36da24/packages/next-auth/src/utils/detect-host.ts#L4

at the time, I didn't want to hunt down other services' environment variables for this, but happy to add support for others out of the box in the new `@auth/*` [framework libraries](https://github.com/nextauthjs/next-auth/blob/1601626d31cc4ad57a33cd7f9fd5c873d4471445/packages/frameworks-sveltekit/src/lib/index.ts#L158).

Happy to take suggestions on how we can make a good/secure default that requires no configuration from the user on popular services! Feel free to open a detailed feature request, and let's leave the derogatory language behind us. This is a community-driven, open source project! 💚

**Reply by DoodlesEpic** [reply]
Thank you very much for this, I'll use that envvar. Maybe add one for render.com which is basically the same as vercel but just `RENDER` instead of `VERCEL`.

Anyways I still think this should not be a feature. If setting an envvar to true just goes around it I don't really get the point of it, what kind of protection does it provide?

Anyways hopefully it will be possible to set `NEXTAUTH_URL` with the current URL (which is already printed to the console nonetheless).

I will look into providing some documentation improvements and feature requests right after Christmas.

Also, merry christmas for you and everyone at Vercel!

Cheers.
